core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java

```diff
@@ -512,29 +525,73 @@ public XContentBuilder rawValue(BytesReference content) throws IOException {
         return this;
     }
 
-    public XContentBuilder timeValueField(String rawFieldName, String readableFieldName, TimeValue timeValue) throws IOException {
-        if (humanReadable) {
-            field(readableFieldName, timeValue.toString());
-        }
-        field(rawFieldName, timeValue.millis());
-        return this;
+    public XContentBuilder field(String name, TimeValue value) throws IOException {
+        return field(name, value, timeUnit());
     }
 
-    public XContentBuilder timeValueField(String rawFieldName, String readableFieldName, long rawTime) throws IOException {
-        if (humanReadable) {
-            field(readableFieldName, new TimeValue(rawTime).toString());
+    public XContentBuilder field(String name, TimeValue value, TimeUnit timeUnit) throws IOException {
+        if (timeUnit == null) {
+            throw new IllegalArgumentException("Time unit cannot be null");
         }
-        field(rawFieldName, rawTime);
-        return this;
-    }
 
-    public XContentBuilder timeValueField(String rawFieldName, String readableFieldName, long rawTime, TimeUnit timeUnit) throws
-        IOException {
         if (humanReadable) {
-            field(readableFieldName, new TimeValue(rawTime, timeUnit).toString());
+            field(name);
+            value(value);
         }
-        field(rawFieldName, rawTime);
-        return this;
+
+        String suffix;
+        Long raw = null;
+        switch (timeUnit) {
+            case NANOSECONDS:
+                suffix = "_in_nanos";
+                if (value != null) {
+                    raw = value.nanos();
+                }
+                break;
+            case MICROSECONDS:
+                suffix = "_in_micros";
+                if (value != null) {
+                    raw = value.micros();
+                }
+                break;
+            case MILLISECONDS:
+                suffix = "_in_millis";
+                if (value != null) {
+                    raw = value.millis();
+                }
+                break;
+            case SECONDS:
+                suffix = "_in_seconds";
+                if (value != null) {
+                    raw = value.seconds();
+                }
+                break;
+            case MINUTES:
+                suffix = "_in_minutes";
+                if (value != null) {
+                    raw = value.minutes();
+                }
+                break;
+            case HOURS:
+                suffix = "_in_hours";
+                if (value != null) {
+                    raw = value.hours();
+                }
+                break;
+            case DAYS:
+                suffix = "_in_days";
+                if (value != null) {
+                    raw = value.days();
+                }
+                break;
```

Comment:
is this switch really needed? Is that because we don't always print out time in millis when non readable?